### ISSUE TITLE
Inaccurate inferred Tuple Dimension

### ISSUE THREAD

**Issue by GBeauregard** [microsoft/pyright#2697]
**Describe the bug**
While programming some shape agnostic matrix code I ran into the following case where pyright infers the wrong size of a tuple leading to a spurious index error.

**To Reproduce**
```python
import itertools
shape = [2, 2]
x = list(itertools.product(*[range(s) for s in shape]))
print(x[0][1])  # Index 1 is out of range for tuple with length 1
reveal_type(x)  # list[tuple[int]]
```

**Expected behavior**
I expected pyright to infer `list[tuple[int, ...]]` instead of `list[tuple[int]]` and not throw an index error.

**VS Code extension or command-line**
pyright commandline Linux 1.1.195

**Additional context**
I checked typeshed and the signature looks okay for itertools.product so I think this is in pyright.

**Comment by erictraut** [comment]
The problem is that `shape` has no known type from the perspective of the type system. Its type is `list[int]`. Likewise, `[range(s) for s in shape]` is of type `list[range]`. The constructor for `product` includes three overloads, but pyright cannot distinguish between them because it doesn't know how many parameters the unpack operator will produce.

If you want this to type check correctly, you'll need to use a different approach. For example, this works:

```python
import itertools
x = list(itertools.product(*(range(2), range(2))))
print(x[0][1])
reveal_type(x)
```

Or if you want to create an internal function that preserves the length of the shape tuple, you could do this:

```python
import itertools
from typing import Tuple, overload

@overload
def create_ranges(shape: Tuple[int]) -> Tuple[range]: ...
@overload
def create_ranges(shape: Tuple[int, int]) -> Tuple[range, range]: ...
@overload
def create_ranges(shape: Tuple[int, int, int]) -> Tuple[range, range, range]: ...
def create_ranges(shape: Tuple[int, ...]) -> Tuple[range, ...]:
    return tuple([range(s) for s in shape])

shape = (2, 2)
x = list(itertools.product(*create_ranges(shape)))
print(x[0][1])
reveal_type(x)
```

**Comment by GBeauregard** [comment]
Wow, this one is tricky. I wasn't expecting a shape problem to come up in untyped code so I'm surprised!

I was expecting pyright to try this overload exactly *because* it doesn't know the shape: https://github.com/python/typeshed/blob/b1f0271da6831c79bc77c3b948b1d808fadaf98b/stdlib/itertools.pyi#L177 (you mentioned 3 overloads - am I looking in the right spot?). Can you elaborate a bit more on what pyright is doing when it's unpacking to match this function's overloads and where this (doesn't) mesh with how pyright normally matches overloads like `*iterables: Iterable[Any]`.

**Comment by erictraut** [comment]
Pyright matches overloads in order, accepting the first one that matches the supplied arguments.

The problem here is that multiple overloads match the supplied argument types in your example. Consider the following, which I think you'd agree should type check without any errors.

```python
def func1(a: int): ...
def func2(a: int, b: int): ...
def func3(*args: int): ...

def func4(*args: int):
    func1(*args)
    func2(*args)
    func3(*args)
```

In the case of all three functions (`func1` through `func3`), `*args` is compatible argument.

Now, let's apply this to the `product.__new__` method. The first overload accepts a single iterable. In other words, it's like `func1` in the example above. When you pass `*args` to this function, the first overload "matches".

An alternative approach to overload matching is to pick the "best" alternative rather than "first" alternative, but that begs the question "what does 'best' mean?". When should the author-supplied ordering of overloads be overruled by the type checker? Unfortunately, PEP 484 isn't clear on this. Experimenting with mypy, it appears that it does employ some form of prioritization in addition to overload ordering.

I'll think about this more. It may be worth experimenting with some heuristics in pyright's overload matching logic.

**Comment by erictraut** [comment]
I've updated the heuristics for overload matching so pyright now picks the overload with a `*args` parameter if the caller passed a `*args` argument. This heuristic overrides the normal "first matching" rule.

**Comment by GBeauregard** [comment]
Nice, and thanks for the explanation; it clarifies specifically what pyright does in overload situations a lot.

**Comment by erictraut** [comment]
This is included in pyright 1.1.196, which I just published. It will also be included in the next release of pylance.
